zpy/client_util.py

```diff
@@ -123,3 +139,259 @@ def clear_last_print():
+    """
+
+    datapoints = []
+    category_count_sums = defaultdict(int)
+
+    # batch level - group images by satapoint
+    for batch in listdir(dataset_path):
+        batch_uri = join(dataset_path, batch)
+        annotation_file_uri = join(batch_uri, "_annotations.zumo.json")
+        metadata = json.load(open(annotation_file_uri))
+
+        for c in values(metadata["categories"]):
+            category_count_sums[c["id"]] += c["count"]
+
+        images_grouped_by_datapoint = group_by(
+            values(metadata["images"]),
+            lambda image: remove_n_extensions(image["relative_path"], n=2),
+        )
+
+        # datapoint level
+        for images in images_grouped_by_datapoint:
+            DATAPOINT_UUID = str(uuid.uuid4())
+            # get [images], [annotations], [categories], {metadata} per data point
+            image_ids = [i["id"] for i in images]
+            annotations = filter_(
+                metadata["annotations"], lambda a: a["image_id"] in image_ids
+            )
+            category_ids = uniq([a["category_id"] for a in annotations])
+            categories = filter_(
+                values(metadata["categories"]), lambda c: c["id"] in category_ids
+            )
+
+            image_new_id_map = {
+                img["id"]: str(
+                    DATAPOINT_UUID
+                    + "-"
+                    + str(Path(img["name"]).suffixes[-2]).replace(".", "")
+                )
+                for img in images
+            }
+
+            # mutate the arrays
+            images_mutated = [
+                {
+                    **i,
+                    "output_path": join(batch_uri, Path(i["relative_path"])),
+                    "id": image_new_id_map[i["id"]],
+                }
+                for i in images
+            ]
+            annotations_mutated = [
+                {
+                    **a,
+                    "image_id": image_new_id_map[a["image_id"]],
+                }
+                for a in annotations
+            ]
+            categories_mutated = [
+                {**c, "count": category_count_sums[c["id"]]} for c in categories
+            ]
+            metadata_mutated = {**metadata["metadata"], "save_path": batch_uri}
+
+            datapoints.append(
+                {
+                    "metadata": metadata_mutated,
+                    "categories": categories_mutated,
+                    "images": images_mutated,
+                    "annotations": annotations_mutated,
+                }
+            )
+
+    def update_category_counts(datapoints):
+        return [
+            {
+                **d,
+                "categories": [
+                    {**c, "count": category_count_sums[c["id"]]}
+                    for c in d["categories"]
+                ],
+            }
+            for d in datapoints
+        ]
+
+    return update_category_counts(datapoints)
```

Comment:
I'd probably just write a comment to give context instead of giving a contextual name to the nested function if that makes sense? I'm not sure if this is a best practices thing or personal preference but I haven't see the nested functions within a function pattern a lot so my hunch is it's better to not do that.

Reply:
Also, this code strikes me as odd and I realized it's because it's duplicating the same category data into every single datapoint. What do you think about changing the function to return a tuple of `(list[datapoint], categories)`? Then in the calling code you'd do something like:
```
    datapoints, categories = group_metadata_by_datapoint(dataset_path)

    if datapoint_callback is not None:
        for datapoint in datapoints:
            datapoint_callback(
                datapoint["images"], datapoint["annotations"], categories
            )
```
I'm not sure, do you think this would be a better design?

Reply:
I think that makes sense, way more sense, and at that point I'd say return a tuple of (metadata, categories, datapoints). No point in having a bunch of duplicates of categories or metadata

Reply:
oh right, the metadata is unique too, yea let's do that!